### PR TITLE
QuantityBox constructor 

### PR DESCRIPTION
**Pull request by shaobinspire**

### CODE REVIEW

**Review by kranar — CHANGES_REQUESTED** [review]
I'm still seeing a lot of these QHash's all over the codebase.

**Comment by kranar** [comment]
Resubmit the pull request once the whole issue has been addressed.

**Comment by shaobinspire** [comment]
I updated the constructors of MoneyBox and PercentBox. Since the specs of Slider and Silder2D don't specify the default modifiers, I left their constructors unchanged.

File: Applications/Spire/Source/Ui/ColorCodePanel.cpp [hunk 1]
```diff
@@ -16,20 +16,13 @@ using namespace Spire;
...
-    return QHash<Qt::KeyboardModifier, T>(
-      {{Qt::NoModifier, 1}, {Qt::AltModifier, 5}, {Qt::ControlModifier, 10},
-        {Qt::ShiftModifier, 20}});
-  }
-
   struct CustomIntegerBox : IntegerBox {
     using EditingFinishSignal = Signal<void (const optional<int>& value)>;
     mutable EditingFinishSignal m_editing_finish_signal;
 
     CustomIntegerBox(std::shared_ptr<OptionalIntegerModel> current,
```
Comment (kranar) [inline Applications/Spire/Source/Ui/ColorCodePanel.cpp:23]: What is this constructed used for? Can you remove it?

**Review by kranar — CHANGES_REQUESTED** [review]

File: Applications/Spire/Source/Ui/ColorCodePanel.cpp (hunk 1 above)
Comment (shaobinspire) [reply Applications/Spire/Source/Ui/ColorCodePanel.cpp]: I don’t think it can be removed. The CustomIntegerBox needs the model of current to construct.

Comment (kranar) [reply Applications/Spire/Source/Ui/ColorCodePanel.cpp]: It needs it for what? The constructor isn't using it for anything.

Comment (shaobinspire) [reply Applications/Spire/Source/Ui/ColorCodePanel.cpp]: It is used to create the RGB color boxes.
```
auto red_box = new CustomIntegerBox(std::move(red_model));
auto green_box = new CustomIntegerBox(std::move(green_model));
auto blue_box = new CustomIntegerBox(std::move(blue_model));
```

Comment (kranar) [reply Applications/Spire/Source/Ui/ColorCodePanel.cpp]: I don't see that happening. Please remove these constructors.

Comment (shaobinspire) [reply Applications/Spire/Source/Ui/ColorCodePanel.cpp]: I am confused. If this constructor is removed, the compiler will show the error of `cannot convert argument 1 from 'std::shared_ptr<Spire::ScalarValueModel<boost::optional<Group>>>' to 'const anonymous-namespace'::CustomIntegerBox &'`.  
My another guess is that you want to remove the custom `IntegerBox`, but I think the custom `IntegerBox` is necessary to implement the specs that the `IntegerBox` displaying nothing can restore to the valid value when it loses focus or when the user presses Enter.

Comment (kranar) [reply Applications/Spire/Source/Ui/ColorCodePanel.cpp]: A constructor definition is only needed if it performs a non-trivial function. If it's not needed then use a default or an inherited constructor:

https://en.cppreference.com/w/cpp/language/using_declaration#Inheriting_constructors

**Review by kranar — APPROVED** [review]
Looks much better.

**Pull request merged**